rtengine/panasonic_decoders.cc

```diff
@@ -33,6 +34,433 @@
+                }
+            } else {
+                for (unsigned col = 0; col < width - 1; col += 2) {
+                    const int c6 = col * 4;
+                    destrow0[col] = srcrow[c6];
+                    destrow0[col + 1] = srcrow[c6 + 2];
+                    destrow1[col] = srcrow[c6 + 4];
+                    destrow1[col + 1] = srcrow[c6 + 6];
+                }
+            }
+        }
+    } catch (...) { // buffer read may throw an exception
+        return false;
+    }
+
+    return true;
+}
+
+uint32_t pana8_param_t::GetDBit(uint64_t a2)
+{
+    for (int i = 0; i < 16; i++) {
+        if ((a2 & hufftable2[i]) == hufftable1[i])
+            return i;
+    }
+
+    return uint32_t((hufftable2[16] & a2) == hufftable1[16]) ^ 0x11u;
+}
+
+pana8_param_t::pana8_param_t(const pana8_tags_t &meta) :
+    gammaTable(0)
+{
+    range_shift = gamma_base = tag3B = 0;
+
+    memset(tag3A, 0, sizeof(tag3A));
+    memset(tag39, 0, sizeof(tag3A));
+    memset(tag3A, 0, sizeof(tag3A));
+    memset(initial, 0, sizeof(tag3A));
+    memset(huff_coeff, 0, sizeof(tag3A));
+    memset(hufftable1, 0, sizeof(tag3A));
+    memset(hufftable2, 0, sizeof(tag3A));
+
+    noGammaFlag = 1;
+
+    for (int i = 0; i < 6; i++) {
+        tag3A[i] = meta.tag3A[i];
+        tag39[i] = meta.tag39[i];
+    }
+
+    tag3B_2 = tag3B = meta.tag3B;
+
+    for (int i = 0; i < 4; i++)
+        initial[i] = meta.initial[i];
+
+    for (int i = 0; i < 17; i++)
+        huff_coeff[i] = (uint32_t(meta.tag41[i]) << 24) | (uint32_t(meta.tag40a[i]) << 16) | meta.tag40b[i];
+
+    std::vector<uint16_t> tempGamma(0x10000);
+    for (unsigned i = 0; i < 0x10000; i++) {
+        uint64_t val = gammaCurve(i);
+        tempGamma[i] = uint16_t(val & 0xffffu);
+        if (i != val)
+            noGammaFlag = 0;
+    }
+
+    if (!noGammaFlag)
+        gammaTable = tempGamma;
+
+    int v7 = 0;
+
+    for (unsigned hindex = 0; hindex < 17; hindex++) {
+        uint32_t hc = huff_coeff[hindex];
+        uint32_t hlow = (hc >> 16) & 0x1F;
+        int16_t v8 = 0;
+        if ((hc & 0x1F0000) != 0) {
+            int h7 = ((hc >> 16) & 0xffffu) & 7;
+            if (hlow - 1 >= 7) {
+                uint32_t hdiff = h7 - hlow;
+                v8 = 0;
+                do {
+                    v8 = (v8 << 8) | 0xFFu;
+                    hdiff += 8;
+                } while (hdiff);
+            } else {
+                v8 = 0;
+            }
+            for (; h7; --h7) {
+                v8 = 2 * v8 + 1;
+            }
+        }
+
+        uint16_t v9 = hc & v8;
+        if (uint32_t(v7) < hlow) {
+            v7 = ((huff_coeff[hindex] >> 16) & 0xFFFFu) & 0x1F;
+        }
+        hufftable2[hindex] = 0xFFFFULL << (64 - hlow);
+        hufftable1[hindex] = (uint64_t)v9 << (64 - hlow);
+    }
+
+    if (v7 < 17) {
+        if (extrahuff.size() < 0x10000)
+            extrahuff.resize(0x10000);
+        uint64_t v17 = 0LL;
+
+        for (int j = 0LL; j != 0x10000; ++j) {
```

Comment:
Do you mean to use `<` instead of `!=` ?

Please note that I just took the libraw code, I don't know how the v8 decoder works and there's no released specification so it's some sort of reverse engineering or something else (looking also at some variable names...)

Reply:
Done